### PR TITLE
Clean up j.u.regex.* and use the native 'd' flag when available

### PR DESCRIPTION
**Pull request by sjrd**
~~Based on #4455.~~

Manually tested with `testSuite2_12/testHtml` in Firefox, which already supports the `'d'` flag by default.

### CODE REVIEW

File: javalib/src/main/scala/java/util/regex/IndicesBuilder.scala [hunk 1]
```diff
@@ -73,31 +78,35 @@ private[regex] class GroupStartMapper private (pattern: String, flags: String,
...
+    // Prepare an `indices` array with the correct length filled with undefined
+    val len = groupCount + 1 // index 0 is the whole match
+    val indices = new IndicesArray(len)
+    indices(0) = js.Tuple2(start, end)
+    var i = 1
     while (i != len) {
-      groupStartMap(i) = -1
+      indices(i) = js.undefined
       i += 1
     }
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/regex/IndicesBuilder.scala:92]: Doesn't a `js.Array` contain `js.undefined` by default?

File: javalib/src/main/scala/java/util/regex/Pattern.scala [hunk 2]
```diff
@@ -134,8 +137,20 @@ final class Pattern private[regex] (
...
+    if (js.isUndefined(lastMatchDyn.indices)) {
+      if (supportsIndices) {
+        if (!enabledNativeIndices) {
+          jsRegExpForFind = new js.RegExp(jsPattern, jsFlagsForFind + "d")
+          jsRegExpForMatches = new js.RegExp(wrapJSPatternForMatches(jsPattern), jsFlags + "d")
+          enabledNativeIndices = true
+        }
+        val regexp = if (forMatches) jsRegExpForMatches else jsRegExpForFind
+        regexp.lastIndex = lastMatch.index
+        lastMatchDyn.indices = regexp.exec(lastMatch.input).asInstanceOf[js.Dynamic].indices
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/regex/Pattern.scala:149]: Consider avoiding the second execution if `lastMatch.indices` is already defined.

Comment (sjrd) [reply javalib/src/main/scala/java/util/regex/Pattern.scala:149]: Hum, I'm not sure I understand. Isn't that what the following line, above, does:
```scala
if (js.isUndefined(lastMatchDyn.indices)) {
```

File: javalib/src/main/scala/java/util/regex/IndicesBuilder.scala (hunk 1 above)
Comment (sjrd) [reply javalib/src/main/scala/java/util/regex/IndicesBuilder.scala:92]: Not really. [It creates an array of empty slots, not of slots with actual `undefined` values.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Array#parameters). That causes the array to become a sparse array, which is less efficient.

Comment (sjrd) [reply javalib/src/main/scala/java/util/regex/IndicesBuilder.scala:99]: I have added a comment about that.

**Review by sjrd — COMMENTED** [review]
Updated.

File: javalib/src/main/scala/java/util/regex/Pattern.scala (hunk 2 above)
Comment (gzm0) [reply javalib/src/main/scala/java/util/regex/Pattern.scala:149]: I don't understand either... Seems fine :P

**Review by gzm0 — APPROVED** [review]

**Pull request merged**